examples/without_gui_qt.py

```diff
@@ -0,0 +1,10 @@
+"""
+Alternative to using napari.gui_qt() context manager. 
```

Comment:
Curious what now happens in the jupyter notebook is `%gui qt` has not been called? Is it same/ different? I just tried it from ipython and everything seemed to magically work without EVER calling `%gui qt`. That surprised me!!! Does this mean we never need to use `%gui qt` anymore?

Reply:
I think this is really core stuff, so forgive the slightly long answer:

### the difference between `run_app()` and `%gui qt`

`%gui qt` does two things (really, the same two things that `napari.gui_qt` does):
1. It [creates an instance of `QGuiApplication` if one doesn't already exist](https://github.com/ipython/ipython/blob/3587f5bb6c8570e7bbb06cf5f7e3bc9b9467355a/IPython/terminal/pt_inputhooks/qt.py#L13-L29).  Which is required to instantiate any Widget... We currently do this in `gui_qt` before the `yield`, but it can really be done anywhere, including _during_ QWidget subclass instantiation (as long as you do it before calling `super().__init__()`)
2. It [creates an event_loop](https://github.com/ipython/ipython/blob/3587f5bb6c8570e7bbb06cf5f7e3bc9b9467355a/IPython/terminal/pt_inputhooks/qt.py#L30), and then basically [enters a cycle](https://github.com/ipython/ipython/blob/3587f5bb6c8570e7bbb06cf5f7e3bc9b9467355a/IPython/terminal/pt_inputhooks/qt.py#L47-L54) of starting that event loop and then quitting it when there is new content ready in the ipython prompt.

Being "in" the event loop means blocking everything in the terminal, or the script, so it's these brief "pauses" of the event loop in ipython that allows you to interact with (for instance) the viewer object in ipython even when the viewer GUI is also visible and interactive (it's basically taking tiny breaks from the event loop to register your input).

To answer your question, what this PR means for us is that:
- no, you don't _need_ to use `%gui qt` or start with `ipython --gui qt` to create a viewer.  (since that just has to do with calling `QApplication([])` once before creating a viewer).
- however, if you don't want your IPython terminal to block when you show the viewer, then having that IPython-created event loop is still handy.  `run_app()` will simply block and wait for you to close the window.  I added a line to `without_gui_qt.py` example to that effect.  The same thing goes for a jupyter notebook:  without %gui qt, cells after `napari.run_app()` won't run until you close the viewer (see the pending cells in this image):

<img width="585" alt="Screen Shot 2021-01-02 at 4 45 36 PM" src="https://user-images.githubusercontent.com/1609449/103467176-6261d800-4d1a-11eb-8de3-ca3ba7402480.png">

on the other hand... with `%gui qt`, the viewer shows and the notebook goes on executing:

<img width="573" alt="Screen Shot 2021-01-02 at 4 46 19 PM" src="https://user-images.githubusercontent.com/1609449/103467179-6e4d9a00-4d1a-11eb-888d-f729bc98a6ed.png">


### why I think we should eventually get rid of gui_qt:

In my opinion, this whole app creation and execution thing isn't really that well suited to a context manager, which usually implies some setup (e.g. "open a file") and teardown (e.g. "close a file").  In our case the "setup" is really a _one-time_ `QApplication` creation... but the "teardown" isn't actually quitting that application, it's just entering the event loop for a moment with `app.exec_()`.  The app instance is _not_ torn down afterwards, *and* the setup isn't even necessary for future viewers ...  which leads to this sort of confusing thing:

```python
# in IPython _without_ having used %gui qt

In [1]: import napari

In [2]: with napari.gui_qt():
   ...:     viewer = napari.Viewer()
   ...:

# good so far the viewer shows, we close it
# ... and we're back to the console.

In [3]: viewer2 = napari.Viewer()  

# wait... why don't I need `with gui_qt()` this time?
# and why don't I see the viewer??

In [4]: with napari.gui_qt():
   ...:     pass
   ...:

# NOW the viewer shows, even though there's _nothing_ in gui_qt
# since all we're really using it for is the `app.exec_` call
# when the context exits.
```

Reply:
ok, yes, I get it now. I missed that `run_app()` was blocking in the notebook, and that `%gui qt` is still useful to avoid that, but blocking in a script is different and so one can now only use `napari.run_app`

You make a very compelling case for dropping `napari.gui_qt`. I think I would prefer to drop it entirely and then just have `napari.run_app` as it is now. I know recently you mentioned `napari.event_loop` as context manager too as an option, but would we still need that too? Was that just an alias of `napari.gui_qt` or was it doing something different that we don't need any more now we have `napari.run_app`?

Reply:
"event_loop" was just a backend agnostic alias for gui_qt.  I think we can probably eventually just use run_app.  Most of the work will just be updating tutorials and stuff. So maybe it's a goal for 0.5.0?

Reply:
Ok yes, brilliant. I'd say we get this in basically as is now for `0.4.3` and then starting `0.4.4` if it has gone well add a deprecation to `gui_qt` saying that it will be removed in `0.5.0` and start moving tutorials/ examples. We can decide details as we get there

Reply:
> Given how important this method will become, how do people feel about just `napari.run()` instead of `napari.run_app()`. I don't think we need the `_app`, as I'm not sure we'll have another `napari.run_` method. In which case I prefer the simpler no `_` version.

yeah, very much open to `napari.run`!  Show matches `plt.show()` ... but I think `.show()` should stay on the viewer itself, rather than `napari.show()`, though in future PRs we can make `viewer.show()` block and run the app depending on the context?

Reply:
Brilliant discussion, thanks @tlambert03 for that long exposition, it clarifies a lot. I agree with deprecating `gui_qt` and removing it in 0.5. 🚀